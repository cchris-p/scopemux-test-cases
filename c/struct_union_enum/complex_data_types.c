/**
 * @file complex_data_types.c
 * @brief Demonstration of structures, unions, and enumerations in C
 *
 * This example shows:
 * - Structure definitions and usage
 * - Nested structures
 * - Unions
 * - Enumerations
 * - Typedef
 * - Bit fields
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

// Simple enumeration
enum Color {
    RED,    // 0 by default
    GREEN,  // 1
    BLUE,   // 2
    YELLOW = 10,
    PURPLE, // 11
    ORANGE  // 12
};

// Enumeration with typedef
typedef enum {
    MONDAY = 1,
    TUESDAY,
    WEDNESDAY,
    THURSDAY,
    FRIDAY,
    SATURDAY,
    SUNDAY
} Weekday;

// Simple structure
struct Point {
    int x;
    int y;
};

// Structure with typedef
typedef struct {
    char title[50];
    char author[50];
    int year;
    float price;
} Book;

// Nested structure
struct Rectangle {
    struct Point topLeft;
    struct Point bottomRight;
};

// Structure with bit fields
typedef struct {
    unsigned int is_active : 1;    // 1 bit
    unsigned int permission : 3;   // 3 bits (0-7)
    unsigned int user_level : 4;   // 4 bits (0-15)
} Flags;

// Union example
union Data {
    int i;
    float f;
    char str[20];
};

// Structure containing a union
typedef struct {
    enum Color color;
    union {
        int i_value;
        float f_value;
        char c_value;
    } value;
} ColoredValue;

/**
 * @brief Print details about a Point structure
 * @param p The Point to print
 */
void print_point(struct Point p) {
    printf("Point(%d, %d)\n", p.x, p.y);
}

/**
 * @brief Calculate the area of a Rectangle
 * @param rect The Rectangle to calculate the area of
 * @return The area of the rectangle
 */
int rectangle_area(struct Rectangle rect) {
    int width = rect.bottomRight.x - rect.topLeft.x;
    int height = rect.bottomRight.y - rect.topLeft.y;
    return width * height;
}

/**
 * @brief Convert a Color enum to its string representation
 * @param color The Color to convert
 * @return String representation of the color
 */
const char* color_to_string(enum Color color) {
    switch (color) {
        case RED:    return "Red";
        case GREEN:  return "Green";
        case BLUE:   return "Blue";
        case YELLOW: return "Yellow";
        case PURPLE: return "Purple";
        case ORANGE: return "Orange";
        default:     return "Unknown";
    }
}

/**
 * @brief Program entry point
 * @return Exit status code
 */
int main() {
    // Using enumerations
    enum Color background = BLUE;
    Weekday today = WEDNESDAY;
    
    printf("Background color: %s (value: %d)\n", color_to_string(background), background);
    printf("Today is day %d of the week\n", today);
    
    // Using structures
    struct Point p1 = {10, 20};
    struct Point p2 = {.y = 5, .x = 15}; // Designated initializers
    
    printf("\nPoints:\n");
    print_point(p1);
    print_point(p2);
    
    // Using nested structures
    struct Rectangle rect = {
        .topLeft = {0, 0},
        .bottomRight = {5, 10}
    };
    
    printf("\nRectangle:\n");
    printf("Top-left: ");
    print_point(rect.topLeft);
    printf("Bottom-right: ");
    print_point(rect.bottomRight);
    printf("Area: %d\n", rectangle_area(rect));
    
    // Using typedef structures
    Book book1;
    strcpy(book1.title, "C Programming");
    strcpy(book1.author, "Dennis Ritchie");
    book1.year = 1978;
    book1.price = 29.99;
    
    printf("\nBook:\n");
    printf("Title: %s\n", book1.title);
    printf("Author: %s\n", book1.author);
    printf("Year: %d\n", book1.year);
    printf("Price: $%.2f\n", book1.price);
    
    // Using bit fields
    Flags user_flags = {1, 5, 10};
    
    printf("\nUser flags:\n");
    printf("Active: %d\n", user_flags.is_active);
    printf("Permission: %d\n", user_flags.permission);
    printf("User level: %d\n", user_flags.user_level);
    
    // Using unions
    union Data data;
    
    data.i = 10;
    printf("\nUnion as int: %d\n", data.i);
    
    data.f = 220.5;
    printf("Union as float: %.1f\n", data.f);
    
    strcpy(data.str, "C Programming");
    printf("Union as string: %s\n", data.str);
    
    // Using structure with union
    ColoredValue cv = {RED, .value.i_value = 100};
    printf("\nColored value:\n");
    printf("Color: %s\n", color_to_string(cv.color));
    printf("Integer value: %d\n", cv.value.i_value);
    
    cv.value.f_value = 99.9;
    printf("Float value (same memory): %f\n", cv.value.f_value);
    
    return 0;
}
