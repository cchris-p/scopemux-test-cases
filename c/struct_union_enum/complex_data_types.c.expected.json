{
  "language": "C",
  "ast": {
    "type": "translation_unit",
    "functions": [
      {
        "name": "print_point",
        "type": "function_definition",
        "return_type": "void",
        "parameters": [
          {"name": "p", "type": "struct Point"}
        ],
        "body_range": {
          "start_line": 75,
          "start_column": 0,
          "end_line": 77,
          "end_column": 1
        },
        "docstring": "Print details about a Point structure\nParam p The Point to print"
      },
      {
        "name": "rectangle_area",
        "type": "function_definition",
        "return_type": "int",
        "parameters": [
          {"name": "rect", "type": "struct Rectangle"}
        ],
        "body_range": {
          "start_line": 85,
          "start_column": 0,
          "end_line": 89,
          "end_column": 1
        },
        "docstring": "Calculate the area of a Rectangle\nParam rect The Rectangle to calculate the area of\nReturn The area of the rectangle"
      },
      {
        "name": "color_to_string",
        "type": "function_definition",
        "return_type": "const char*",
        "parameters": [
          {"name": "color", "type": "enum Color"}
        ],
        "body_range": {
          "start_line": 97,
          "start_column": 0,
          "end_line": 107,
          "end_column": 1
        },
        "docstring": "Convert a Color enum to its string representation\nParam color The Color to convert\nReturn String representation of the color"
      },
      {
        "name": "main",
        "type": "function_definition",
        "return_type": "int",
        "parameters": [],
        "body_range": {
          "start_line": 113,
          "start_column": 0,
          "end_line": 178,
          "end_column": 1
        },
        "docstring": "Program entry point\nReturn Exit status code"
      }
    ],
    "includes": [
      {
        "path": "stdio.h",
        "system": true,
        "line": 12
      },
      {
        "path": "string.h",
        "system": true,
        "line": 13
      },
      {
        "path": "stdlib.h",
        "system": true,
        "line": 14
      }
    ],
    "comments": [
      {
        "type": "block_comment",
        "text": "This example shows:\n- Structure definitions and usage\n- Nested structures\n- Unions\n- Enumerations\n- Typedef\n- Bit fields",
        "range": {
          "start_line": 1,
          "start_column": 0,
          "end_line": 10,
          "end_column": 3
        }
      },
      {
        "type": "line_comment",
        "text": "Simple enumeration",
        "line": 16
      },
      {
        "type": "line_comment",
        "text": "0 by default",
        "line": 17
      },
      {
        "type": "line_comment",
        "text": "1",
        "line": 18
      },
      {
        "type": "line_comment",
        "text": "2",
        "line": 19
      },
      {
        "type": "line_comment",
        "text": "11",
        "line": 21
      },
      {
        "type": "line_comment",
        "text": "12",
        "line": 22
      },
      {
        "type": "line_comment",
        "text": "Enumeration with typedef",
        "line": 25
      },
      {
        "type": "line_comment",
        "text": "Simple structure",
        "line": 35
      },
      {
        "type": "line_comment",
        "text": "Structure with typedef",
        "line": 40
      },
      {
        "type": "line_comment",
        "text": "Nested structure",
        "line": 48
      },
      {
        "type": "line_comment",
        "text": "Structure with bit fields",
        "line": 54
      },
      {
        "type": "line_comment",
        "text": "1 bit",
        "line": 55
      },
      {
        "type": "line_comment",
        "text": "3 bits (0-7)",
        "line": 56
      },
      {
        "type": "line_comment",
        "text": "4 bits (0-15)",
        "line": 57
      },
      {
        "type": "line_comment",
        "text": "Union example",
        "line": 61
      },
      {
        "type": "line_comment",
        "text": "Structure containing a union",
        "line": 68
      },
      {
        "type": "block_comment",
        "text": "Print details about a Point structure\nParam p The Point to print",
        "range": {
          "start_line": 72,
          "start_column": 0,
          "end_line": 75,
          "end_column": 3
        }
      },
      {
        "type": "block_comment",
        "text": "Calculate the area of a Rectangle\nParam rect The Rectangle to calculate the area of\nReturn The area of the rectangle",
        "range": {
          "start_line": 81,
          "start_column": 0,
          "end_line": 85,
          "end_column": 3
        }
      },
      {
        "type": "block_comment",
        "text": "Convert a Color enum to its string representation\nParam color The Color to convert\nReturn String representation of the color",
        "range": {
          "start_line": 93,
          "start_column": 0,
          "end_line": 97,
          "end_column": 3
        }
      },
      {
        "type": "block_comment",
        "text": "Program entry point\nReturn Exit status code",
        "range": {
          "start_line": 110,
          "start_column": 0,
          "end_line": 113,
          "end_column": 3
        }
      },
      {
        "type": "line_comment",
        "text": "Using enumerations",
        "line": 114
      },
      {
        "type": "line_comment",
        "text": "Using structures",
        "line": 120
      },
      {
        "type": "line_comment",
        "text": "Designated initializers",
        "line": 121
      },
      {
        "type": "line_comment",
        "text": "Using nested structures",
        "line": 127
      },
      {
        "type": "line_comment",
        "text": "Using typedef structures",
        "line": 138
      },
      {
        "type": "line_comment",
        "text": "Using bit fields",
        "line": 149
      },
      {
        "type": "line_comment",
        "text": "Using unions",
        "line": 156
      },
      {
        "type": "line_comment",
        "text": "Using structure with union",
        "line": 167
      }
    ],
    "enums": [
      {
        "name": "Color",
        "type": "enum_definition",
        "values": [
          {"name": "RED", "value": 0},
          {"name": "GREEN", "value": 1},
          {"name": "BLUE", "value": 2},
          {"name": "YELLOW", "value": 10},
          {"name": "PURPLE", "value": 11},
          {"name": "ORANGE", "value": 12}
        ],
        "range": {
          "start_line": 16,
          "start_column": 0,
          "end_line": 23,
          "end_column": 1
        }
      },
      {
        "name": "Weekday",
        "type": "enum_definition",
        "typedef": true,
        "values": [
          {"name": "MONDAY", "value": 1},
          {"name": "TUESDAY", "value": 2},
          {"name": "WEDNESDAY", "value": 3},
          {"name": "THURSDAY", "value": 4},
          {"name": "FRIDAY", "value": 5},
          {"name": "SATURDAY", "value": 6},
          {"name": "SUNDAY", "value": 7}
        ],
        "range": {
          "start_line": 25,
          "start_column": 0,
          "end_line": 33,
          "end_column": 1
        }
      }
    ],
    "structs": [
      {
        "name": "Point",
        "type": "struct_definition",
        "members": [
          {"name": "x", "type": "int"},
          {"name": "y", "type": "int"}
        ],
        "range": {
          "start_line": 35,
          "start_column": 0,
          "end_line": 38,
          "end_column": 1
        }
      },
      {
        "name": "Book",
        "type": "struct_definition",
        "typedef": true,
        "members": [
          {"name": "title", "type": "char[50]"},
          {"name": "author", "type": "char[50]"},
          {"name": "year", "type": "int"},
          {"name": "price", "type": "float"}
        ],
        "range": {
          "start_line": 40,
          "start_column": 0,
          "end_line": 46,
          "end_column": 1
        }
      },
      {
        "name": "Rectangle",
        "type": "struct_definition",
        "members": [
          {"name": "topLeft", "type": "struct Point"},
          {"name": "bottomRight", "type": "struct Point"}
        ],
        "range": {
          "start_line": 48,
          "start_column": 0,
          "end_line": 51,
          "end_column": 1
        }
      },
      {
        "name": "Flags",
        "type": "struct_definition",
        "typedef": true,
        "members": [
          {"name": "is_active", "type": "unsigned int", "bit_field": 1},
          {"name": "permission", "type": "unsigned int", "bit_field": 3},
          {"name": "user_level", "type": "unsigned int", "bit_field": 4}
        ],
        "range": {
          "start_line": 54,
          "start_column": 0,
          "end_line": 58,
          "end_column": 1
        }
      },
      {
        "name": "ColoredValue",
        "type": "struct_definition",
        "typedef": true,
        "members": [
          {"name": "color", "type": "enum Color"},
          {
            "name": "value",
            "type": "union",
            "members": [
              {"name": "i_value", "type": "int"},
              {"name": "f_value", "type": "float"},
              {"name": "c_value", "type": "char"}
            ]
          }
        ],
        "range": {
          "start_line": 68,
          "start_column": 0,
          "end_line": 75,
          "end_column": 1
        }
      }
    ],
    "unions": [
      {
        "name": "Data",
        "type": "union_definition",
        "members": [
          {"name": "i", "type": "int"},
          {"name": "f", "type": "float"},
          {"name": "str", "type": "char[20]"}
        ],
        "range": {
          "start_line": 61,
          "start_column": 0,
          "end_line": 65,
          "end_column": 1
        }
      }
    ],
    "variable_declarations": [
      {
        "name": "background",
        "type": "enum Color",
        "line": 114,
        "initialized": true
      },
      {
        "name": "today",
        "type": "Weekday",
        "line": 115,
        "initialized": true
      },
      {
        "name": "p1",
        "type": "struct Point",
        "line": 120,
        "initialized": true
      },
      {
        "name": "p2",
        "type": "struct Point",
        "line": 121,
        "initialized": true
      },
      {
        "name": "rect",
        "type": "struct Rectangle",
        "line": 127,
        "initialized": true
      },
      {
        "name": "book1",
        "type": "Book",
        "line": 138,
        "initialized": false
      },
      {
        "name": "user_flags",
        "type": "Flags",
        "line": 149,
        "initialized": true
      },
      {
        "name": "data",
        "type": "union Data",
        "line": 156,
        "initialized": false
      },
      {
        "name": "cv",
        "type": "ColoredValue",
        "line": 167,
        "initialized": true
      }
    ],
    "function_calls": [
      {
        "name": "printf",
        "arguments": [
          "\"Point(%d, %d)\\n\"",
          "p.x",
          "p.y"
        ],
        "line": 76
      },
      {
        "name": "printf",
        "arguments": [
          "\"Background color: %s (value: %d)\\n\"",
          "color_to_string(background)",
          "background"
        ],
        "line": 117
      },
      {
        "name": "color_to_string",
        "arguments": [
          "background"
        ],
        "line": 117
      },
      {
        "name": "printf",
        "arguments": [
          "\"Today is day %d of the week\\n\"",
          "today"
        ],
        "line": 118
      },
      {
        "name": "printf",
        "arguments": [
          "\"\\nPoints:\\n\""
        ],
        "line": 123
      },
      {
        "name": "print_point",
        "arguments": [
          "p1"
        ],
        "line": 124
      },
      {
        "name": "print_point",
        "arguments": [
          "p2"
        ],
        "line": 125
      },
      {
        "name": "printf",
        "arguments": [
          "\"\\nRectangle:\\n\""
        ],
        "line": 132
      },
      {
        "name": "printf",
        "arguments": [
          "\"Top-left: \""
        ],
        "line": 133
      },
      {
        "name": "print_point",
        "arguments": [
          "rect.topLeft"
        ],
        "line": 134
      },
      {
        "name": "printf",
        "arguments": [
          "\"Bottom-right: \""
        ],
        "line": 135
      },
      {
        "name": "print_point",
        "arguments": [
          "rect.bottomRight"
        ],
        "line": 136
      },
      {
        "name": "printf",
        "arguments": [
          "\"Area: %d\\n\"",
          "rectangle_area(rect)"
        ],
        "line": 137
      },
      {
        "name": "rectangle_area",
        "arguments": [
          "rect"
        ],
        "line": 137
      },
      {
        "name": "strcpy",
        "arguments": [
          "book1.title",
          "\"C Programming\""
        ],
        "line": 139
      },
      {
        "name": "strcpy",
        "arguments": [
          "book1.author",
          "\"Dennis Ritchie\""
        ],
        "line": 140
      },
      {
        "name": "printf",
        "arguments": [
          "\"\\nBook:\\n\""
        ],
        "line": 144
      },
      {
        "name": "printf",
        "arguments": [
          "\"Title: %s\\n\"",
          "book1.title"
        ],
        "line": 145
      },
      {
        "name": "printf",
        "arguments": [
          "\"Author: %s\\n\"",
          "book1.author"
        ],
        "line": 146
      },
      {
        "name": "printf",
        "arguments": [
          "\"Year: %d\\n\"",
          "book1.year"
        ],
        "line": 147
      },
      {
        "name": "printf",
        "arguments": [
          "\"Price: $%.2f\\n\"",
          "book1.price"
        ],
        "line": 148
      },
      {
        "name": "printf",
        "arguments": [
          "\"\\nUser flags:\\n\""
        ],
        "line": 151
      },
      {
        "name": "printf",
        "arguments": [
          "\"Active: %d\\n\"",
          "user_flags.is_active"
        ],
        "line": 152
      },
      {
        "name": "printf",
        "arguments": [
          "\"Permission: %d\\n\"",
          "user_flags.permission"
        ],
        "line": 153
      },
      {
        "name": "printf",
        "arguments": [
          "\"User level: %d\\n\"",
          "user_flags.user_level"
        ],
        "line": 154
      },
      {
        "name": "printf",
        "arguments": [
          "\"\\nUnion as int: %d\\n\"",
          "data.i"
        ],
        "line": 159
      },
      {
        "name": "printf",
        "arguments": [
          "\"Union as float: %.1f\\n\"",
          "data.f"
        ],
        "line": 162
      },
      {
        "name": "strcpy",
        "arguments": [
          "data.str",
          "\"C Programming\""
        ],
        "line": 164
      },
      {
        "name": "printf",
        "arguments": [
          "\"Union as string: %s\\n\"",
          "data.str"
        ],
        "line": 165
      },
      {
        "name": "printf",
        "arguments": [
          "\"\\nColored value:\\n\""
        ],
        "line": 168
      },
      {
        "name": "printf",
        "arguments": [
          "\"Color: %s\\n\"",
          "color_to_string(cv.color)"
        ],
        "line": 169
      },
      {
        "name": "color_to_string",
        "arguments": [
          "cv.color"
        ],
        "line": 169
      },
      {
        "name": "printf",
        "arguments": [
          "\"Integer value: %d\\n\"",
          "cv.value.i_value"
        ],
        "line": 170
      },
      {
        "name": "printf",
        "arguments": [
          "\"Float value (same memory): %f\\n\"",
          "cv.value.f_value"
        ],
        "line": 173
      }
    ]
  }
}
